### PR TITLE
Add support for USING to SQL unparser

### LINKED ISSUE #10652: Support convert LogicalPlan JOIN with `Using` constraint to SQL String
### Is your feature request related to a problem or challenge?

We only support to convert JOIN with `ON` constraint to SQL String now. The SQL as below can't be converted now.
```
SELECT * FROM orders JOIN customer USING (custkey)
```
It will panic with the message:
```
NotImplemented("Unsupported join constraint: Using")
```

### Describe the solution you'd like

The unparse rule for `JOIN` in `datafusion/sql/src/unparser/plan.rs` should be implemented.
https://github.com/apache/datafusion/blob/8bedecc00b2f1f04d7b1a907152ce0d19b7046a5/datafusion/sql/src/unparser/plan.rs#L277-L285

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by wackywendell**
## Which issue does this PR close?

Closes #10652.

## Are these changes tested?

Yes - I added a couple example cases to the existing tests.

## Are there any user-facing changes?

Only the listed change - `LogicalPlan`s with `JoinConstraint::Using` can now successfully be translated back to SQL.

## Other

👋  Hello! This is my first contribution to apache/datafusion; all feedback welcome! I expect there will be a bit of learning on my side as to how this community works. Thanks for any help there you can provide!

Also - this passes `cargo test` locally, but if someone could trigger CI please, that would be appreciated 🙇

### CODE REVIEW

File: datafusion/sql/src/unparser/plan.rs [hunk 1]
```diff
@@ -582,9 +552,62 @@ impl Unparser<'_> {
...
+    /// Convert the components of a USING clause to the USING AST
+    fn join_using_to_sql(&self, join_conditions: &[(Expr, Expr)]) -> Result<ast::JoinConstraint> {
+        let mut idents = Vec::with_capacity(join_conditions.len());
+        for (left, right) in join_conditions {
+            match (left, right) {
+                (Expr::Column(Column { relation: _, name: left_name }), Expr::Column(Column { relation: _, name: right_name })) => {
+                    if left_name != right_name {
+                        // USING is only valid when the column names are the
+                        // same, so they should never be different
+                        return not_impl_err!("Unsupported USING with different column names");
```
Comment (wackywendell) [inline datafusion/sql/src/unparser/plan.rs:564]: Should these be `not_impl_err`? If I understand correctly, this is an invalid state, so arguably its something like an `Internal` error... same for below.

**Comment by wackywendell** [comment]
^ I installed the `pre-commit.sh` and turned on clippy and `cargo fmt` in my editor; caught the above in 11324895d04.

Speaking of which - should I squash/rebase? By default I tend not to do that on public PRs/branches, as it can be confusing for reviews.

File: datafusion/sql/src/unparser/plan.rs [hunk 2]
```diff
@@ -582,9 +558,89 @@ impl Unparser<'_> {
...
+                        relation: _,
+                        name: right_name,
+                    }),
+                ) => {
+                    if left_name != right_name {
+                        // USING is only valid when the column names are the
+                        // same, so they should never be different
+                        return not_impl_err!(
+                            "Unsupported USING with different column names"
+                        );
```
Comment (goldmedal) [inline datafusion/sql/src/unparser/plan.rs:584]: IMO, `not_impl_err` is for some todo items. If this case never happens, I prefer to use `internal_err` here.
```suggestion
                        return internal_err!(
                            "Unsupported USING with different column names"
                        );
```

**Review by goldmedal — APPROVED** [review]
Thanks @wackywendell for working on this. Overall LGTM. 
However, I can't help you to trigger the CI. Could @alamb or @jayzhan211 help to trigger it?

**Comment by goldmedal** [comment]
> Speaking of which - should I squash/rebase? By default I tend not to do that on public PRs/branches, as it can be confusing for reviews.

Unless you think that rebase or squash makes the review easier, I don't think you need to do it. We don't have any rules requiring you to do so.

**Review by alamb — COMMENTED** [review]
Thanks @wackywendell and @goldmedal  -- this is a great first PR. I left some comments

File: datafusion/sql/src/unparser/plan.rs (hunk 2 above)
Comment (wackywendell) [reply datafusion/sql/src/unparser/plan.rs:584]: That can happen with `ON`, but it can't happen with `USING`, right? This function is only for `USING`.

Comment (wackywendell) [reply datafusion/sql/src/unparser/plan.rs:584]: So I thought about this some more - these two cases should be impossible for a `LogicalPlan` made from SQL (you can't do `JOIN … USING (t1.a = t1.b)`, but could be created by a user directly creating a `LogicalPlan::Join(Join { … })`.

Given that, fallback makes sense, and I'll add a comment to explain.

**Comment by wackywendell** [comment]
Thanks all for your super-quick reviews! Working on an update now.

**Comment by alamb** [comment]
> Unless you think that rebase or squash makes the review easier, I don't think you need to do it. We don't have any rules requiring you to do so.

Agreed -- the final merge into main is done with squash (so the entire PR turns into a single commit). I typically find it easier to review PRs where subsequent changes are made with new commits (rather than rebases) so it is more clear what changed

**Comment by wackywendell** [comment]
Pushed an update - to get the fallback from `USING` to `ON` to work and be reasonably succinct, I refactored a bit. I think it will need another review now - thanks so much!

**Review by alamb — APPROVED** [review]
Thanks again

**Comment by alamb** [comment]
🚀

**Pull request merged**